src/app.js

```diff
@@ -125,20 +127,46 @@ gpii.app.keyIn = function (token) {
   * Keys out of the GPII.
   * Currently uses an url to key out although this should be changed to use Electron IPC.
   * @param token {String} The token to key out with.
+  * @return {Promise} A promise that will be resolved/rejected when the request is finished.
   */
 gpii.app.keyOut = function (token) {
-    request("http://localhost:8081/user/" + token + "/logout", function (/*error, response, body*/) {
+    var togo = fluid.promise();
+    request("http://localhost:8081/user/" + token + "/logout", function (error, response, body) {
         //TODO Put in some error logging
+        if (error) {
+            togo.reject(error);
+        } else {
+            togo.resolve();
+        }
     });
+    return togo;
 };
 
 /**
   * Stops the Electron Application.
+  * @return {Promise} An already resolved promise.
   */
-gpii.app.exit = function () {
-    //TODO: This should stop the GPII gracefully before quitting the application.
+gpii.app.performQuit = function () {
     var app = require("electron").app;
+    var togo = fluid.promise();
+
+    gpii.stop();
     app.quit();
+
+    togo.resolve();
+    return togo;
+}
+
+/**
+  * Handles the exit of the Electron Application.
+  * @param that {Component} An instance of gpii.app
+  */
+gpii.app.exit = function (that) {
+    if (that.model.keyedInUserToken) {
+        fluid.promise.sequence([that.keyOut(that.model.keyedInUserToken), gpii.app.performQuit]);
```

Comment:
Now you have a failure handler in keyOut, it is now possible that failure of the keyout request can abort the quit action. You should either wrap the the request promise using a utility such as this one, https://github.com/GPII/universal/blob/master/gpii/node_modules/lifecycleManager/src/LifecycleManager.js#L353 , or else put similar code inside keyOut itself.